app/src/main/java/org/mozilla/focus/home/HomeFragment.kt

```diff
@@ -90,6 +92,26 @@ class HomeFragment : Fragment() {
         }
 
         registerForContextMenu(view)
+
+        tileContainer.addOnScrollListener(object : RecyclerView.OnScrollListener() {
+            override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
+                // Only add scroll offset when user is scrolling down
+                if (dy > 0) {
+                    val layoutManager = recyclerView?.layoutManager as GridLayoutManager
+                    val visibleItemCount = layoutManager.childCount
+                    val totalItemCount = layoutManager.itemCount
+                    val pastVisibleItems = layoutManager.findFirstVisibleItemPosition()
+                    // This checks to see if we have reached the last row of the RecyclerView after a scroll occurs.
+                    // If we have not reached that point, we additionally scroll with a small offset
+                    // to reveal the next row, thus preventing the refocusing of the URL bar.
+                    if (pastVisibleItems + visibleItemCount < totalItemCount) {
```

Comment:
If we have more than three rows (say 4), won't this scroll extra when you move down to the second row, even if you haven't reached the "bottom" of the visible items? because
`pastVisibleItems` = 0
`visibleItemCount` = 3 x 5 = 15
`totalItemCount` = 4 x 5 = 20 > 15

Reply:
I think we want to try to scroll extra either when we reach the bottom "visible" row, or when we reach a certain margin from the bottom.

Reply:
No it won't because onScrolled only gets called when you move down to the third row, the view has to physically scroll for the function to be called, not the focus to be changed.

Reply:
So as long as there are still items left underneath what was just scrolled, an extra scroll is needed, that's how I thought about it

Reply:
Okay, I see, that makes sense. So one problem I see is that a scroll action will trigger more nested scroll actions because this is called on every scroll, so the view will just autoscroll until it reaches the bottom. This isn't a problem if you don't have that many sites (or just one more row).

Reply:
Since we're not in a rush to land, I think we should try to take care of edge cases. If you come up with a fix for this edge case, please push it as a *separate commit*.